lib/spack/spack/directives.py

```diff
@@ -465,23 +465,50 @@ def variant(
         multi (bool): if False only one value per spec is allowed for
             this variant
         validator (callable): optional group validator to enforce additional
-            logic. It receives a tuple of values and should raise an instance
-            of SpackError if the group doesn't meet the additional constraints
+            logic. It receives the package name, the variant name and a tuple
+            of values and should raise an instance of SpackError if the group
+            doesn't meet the additional constraints
+
+    Raises:
+        DirectiveError: if arguments passed to the directive are invalid
     """
     if name in reserved_names:
-        raise ValueError("The variant name '%s' is reserved by Spack" % name)
+        raise DirectiveError(
+            "The variant name '%s' is reserved by Spack" % name
+        )
 
+    # Ensure we have a sequence of allowed variant values, or a
+    # predicate for it.
     if values is None:
-        if default in (True, False) or (type(default) is str and
-                                        default.upper() in ('TRUE', 'FALSE')):
+        if str(default).upper() in ('TRUE', 'FALSE'):
             values = (True, False)
         else:
             values = lambda x: True
 
+    # The object defining variant values might supply its own defaults for
+    # all the other arguments. Ensure we have no conflicting definitions
+    # in place.
+    for argument in ('default', 'multi', 'validator'):
```

Comment:
> Associating these functions under some common namespace like multi would also help to clarify things.

I agree with changing the error message but not with this point. In this PR I am using a variant function for a multi-valued variant but the implementation is such that we could use it for any kind of variant in the future, should the need arise. I'd like to maintain this flexibility, as it comes for free. Would the message:
```
cannot specify "{0}" when passing a variant object to the argument "values" of a variant directive [variant "{name}", package "{pkg.name}"]
```
be fine with you?

> I think a user would be confused by this message, since the variant functions implicitly define these values. 

Just to clarify a thing, in case it's needed. This error is not meant for a normal user (a guy that just want to install software with Spack) as a package failing in this loop won't pass the `test_package_sanity` unit test in Travis. It will be seen only by packagers.

Reply:
> In this PR I am using a variant function for a multi-valued variant but the implementation is such that we could use it for any kind of variant in the future

On one hand I get what you mean. On the other hand in the future this could be refactored such that it receives objects that set `multi`, `default`, and `validator` so you can trace back to the class generating the conflict. To be clear I don't think you would have to handle that now - a TODO would be sufficient.

> a package failing in this loop won't pass the test_package_sanity unit test in Travis. It will be seen only by packagers.

That's who I had in mind when I complained about the error message. When possible I want to reduce the level of sophistication required for writing packages. Also if a package writer hits this issue I'm pretty sure they'd end up opening an issue, and it took me a few minutes from looking at the code to understand what was actually going on.

Reply:
Minor modification to your suggestion:
`Remove specification of {0} argument: it is handled by an attribute of the "values" argument`